docs/userguide/apps.rst

```diff
@@ -1,27 +1,61 @@
 
-Parsl apps are defined by annotating Python functions with an app decorator. Currently two types of apps can be defined: Python, with the corresponding ``@python_app`` decorator, and Bash, with the corresponding ``@bash_app`` decorator. Python apps encapsulate pure Python code, while Bash apps wrap calls to external applications and scripts.
+A Parsl app is defined by annotating a Python function with an app decorator. 
+Currently two types of apps can be defined: a **Python app**, with the ``@python_app`` decorator, and a **Bash app**, with the ``@bash_app`` decorator. 
+Python apps encapsulate pure Python code, while Bash apps wrap calls to external applications and scripts.
 
 Python Apps
 -----------
 
-The following code snippet shows a Python function ``double(x: int)``, used to double the input value. This function is defined as a Parsl app using the ``@python_app`` decorator.
-
-Python apps are *pure* Python functions. As these functions are executed asynchronously, and potentially remotely, it is important to note that they must explicitly import any required modules and act only on defined input arguments (i.e., they cannot include variables used elsewhere in the script).
+The following code snippet shows a Python function ``double(x: int)``, used to double the value provided as input. 
+The ``@python_app`` decorator defines the function as a Parsl Python app.  
 
 .. code-block:: python
 
        @python_app
        def double(x):
              return x * 2
 
-       double(x)
+       double(42)
+
+As a Parsl Python app is executed asynchronously, and potentially remotely, it must explicitly import any required modules and cannot refer to variables used outside the function. 
+Thus while the following code fragment is valid Python, it is not valid Parsl, as the `bad_double()` function requires the model `random` and refers to the external variable `factor`.
+
+.. code-block:: python
+
+       import random
+       factor = 5
+
+       @python_app
+       def bad_double(x):
+             return x * random.random() * factor
+
+       print(bad_double(42))
+       
+The following alternative formulation is valid Parsl.
+
+.. code-block:: python
+
+       import random
+       factor = 5
 
-Python apps may also act upon files. In order to make Parsl aware of these files they must be defined using the ``inputs`` or ``outputs`` keyword arguments. The following code snippet illustrates how the contents of one file can be copied to another.
+       @python_app
+       def good_double(x, f):
+             import random
+             return x * random.random() * f
+
+       print(good_double(42, factor))
+
+An input argument 
+may be a Python object or a DataFuture (see :ref:`label-futures`) returned by another app. 
```

Comment:
Input arguments could be regular python objects or futures (AppFuture or DataFuture) returned by apps.

Do you think we should clarify here that Futures could be passed as args or kwargs, however we do not inspect nested data structures that may contain futures ?